rust-runtime/aws-smithy-http-server/src/plugin/check_health.rs

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Middleware for handling health check requests.
+//!
+//! # Example
+//!
+//! ```rust
+//! # use aws_smithy_http_server::{body, plugin::{PluginPipeline, check_health::CheckHealthLayer}};
+//! # use hyper::{Body, Response, StatusCode};
+//! let plugins = PluginPipeline::new()
+//!     // Handle all `/ping` health check requests by returning a `200 OK`.
+//!     .http_layer(CheckHealthLayer::new("/ping", |_req| async {
+//!         Response::builder()
+//!             .status(StatusCode::OK)
+//!             .body(body::boxed(Body::empty()))
+//!             .expect("Couldn't construct response")
+//!     }));
+//!
+//! ```
+
+use std::task::{Context, Poll};
+
+use futures_util::Future;
+use hyper::{Body, Request, Response};
+use pin_project_lite::pin_project;
+use tower::{util::Oneshot, Layer, Service, ServiceExt};
+
+use crate::body::BoxBody;
+
+use super::either::EitherProj;
+use super::Either;
+
+/// A [`tower::Layer`] used to apply [`CheckHealthService`].
+#[derive(Clone, Debug)]
+pub struct CheckHealthLayer<PingHandler> {
+    health_check_uri: &'static str,
+    ping_handler: PingHandler,
+}
+
+impl CheckHealthLayer<()> {
+    pub fn new<HandlerFuture: Future<Output = Response<BoxBody>>, H: Fn(Request<Body>) -> HandlerFuture>(
+        health_check_uri: &'static str,
+        ping_handler: H,
+    ) -> CheckHealthLayer<H> {
+        CheckHealthLayer {
+            health_check_uri,
+            ping_handler,
+        }
+    }
+}
+
+impl<S, H: Clone> Layer<S> for CheckHealthLayer<H> {
+    type Service = CheckHealthService<H, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CheckHealthService {
+            inner,
+            layer: self.clone(),
+        }
+    }
+}
+
+/// A middleware [`Service`] responsible for handling health check requests.
+#[derive(Clone, Debug)]
+pub struct CheckHealthService<H, S> {
+    inner: S,
+    layer: CheckHealthLayer<H>,
+}
+
+impl<H, HandlerFuture, S> Service<Request<Body>> for CheckHealthService<H, S>
+where
+    S: Service<Request<Body>, Response = Response<BoxBody>> + Clone,
+    S::Future: std::marker::Send + 'static,
+    HandlerFuture: Future<Output = Response<BoxBody>>,
+    H: Fn(Request<Body>) -> HandlerFuture,
```

Comment:
Why restrict the ping handler to an async function? Why not an arbitrary full-blown `Service`?

Reply:
Two constructors would be nice, using `service_fn` in the case where we pass a handler.

Reply:
Given that this is going to be `ALB`-specific, I'm not sure we need to allow a full blown service in here, even this closure is too general: AFAIK, ALB only cares about the status code of the response you send, so maybe we would just need a `Fn() -> StatusCode` here?

Reply:
We could make the check just be a determination of a status code if that's all that ALB cares about; the interface would model a tighter reality and more apparent to the user as to what really matters. I like it.

As to what we should take in here, my thoughts are: healthchecks should be fast and simple, e.g. verifying some in-memory state or ensuring the filesystem is ok. An `async` context is thus probably preferred, since, despite it allowing the user to possibly do some dumb costly stuff, filesystem operations are async. It seems a bit arbitrary to restrict it to be an async function here though, since our composition unit in the framework has been the Tower service. I expect almost all healthchecks to be implemented as simple async functions, so I think having a constructor for `AlbHealthCheckLayer` that takes in an async function is good. But it'd also be neat to have a second constructor take in a service; we can use `service_fn` to lift the function in the former constructor and continue working in the library code with just Tower services like Harry suggests. The Tower service can have `type Response = http::StatusCode` too, it doesn't have to return HTTP responses.

Reply:
Done in https://github.com/awslabs/smithy-rs/pull/2540/commits/f22e0129e9ae88e0447b525132a8105804d1343d.
I'm not happy with how complex the type of `new_fn` is, but anything simpler than that caused the pokemon service example to fail to compile :(.